protocol-units/da-sequencer/node/src/celestia/submit.rs

```diff
@@ -0,0 +1,116 @@
+//! The blob submitter task.
+
+use super::{CelestiaBlobData, CelestiaHeight, ExternalDaNotification};
+use crate::block::{BlockHeight, SequencerBlockDigest};
+
+use anyhow::{anyhow, Context};
+use celestia_rpc::prelude::*;
+use celestia_rpc::{Client, TxConfig};
+use celestia_types::nmt::Namespace;
+use celestia_types::{AppVersion, Blob};
+use tokio::select;
+use tokio::sync::mpsc;
+use tracing::debug;
+
+use std::mem;
+use std::sync::Arc;
+
+// Size limit to ensure inclusion in a Celestia block, as per
+// https://docs.celestia.org/how-to-guides/submit-data#maximum-blob-size
+const MAX_CELESTIA_BLOB_SIZE: usize = 512 * 1024;
+
+pub(crate) struct BlobSubmitter {
+	// The Celestia RPC client
+	celestia_client: Arc<Client>,
+	// The Celestia namespace
+	celestia_namespace: Namespace,
+	// Channel to receive digests from foreground
+	digest_receiver: mpsc::Receiver<SequencerBlockDigest>,
+	// Channel to send notifications from Celestia layer
+	notifier: mpsc::Sender<ExternalDaNotification>,
+}
+
+impl BlobSubmitter {
+	pub(crate) fn new(
+		celestia_client: Arc<Client>,
+		celestia_namespace: Namespace,
+		digest_receiver: mpsc::Receiver<SequencerBlockDigest>,
+		notifier: mpsc::Sender<ExternalDaNotification>,
+	) -> Self {
+		BlobSubmitter { celestia_client, celestia_namespace, digest_receiver, notifier }
+	}
+
+	pub(crate) async fn run(mut self) -> Result<(), anyhow::Error> {
+		// Digests accumulated while waiting for client to submit
+		let mut buffered_digests: Vec<SequencerBlockDigest> = vec![];
+		// Size of the accumulated blob data
+		let mut total_data_size = 0;
+		let mut submit_request = None;
+		loop {
+			match &mut submit_request {
+				None => {
+					// No request is currently pending.
+					// Grab the accumulated digests, if there are any, and submit them in a blob.
+					if !buffered_digests.is_empty() {
+						let digests = mem::replace(&mut buffered_digests, vec![]);
+						submit_request = Some(Box::pin(submit_blob(
+							&self.celestia_client,
+							self.celestia_namespace.clone(),
+							digests,
+						)));
+					}
+				}
+				Some(pending_request) => {
+					// While a submit request is pending, accumulate blobs.
+					// Provide back-pressure if the data size is pushing
+					// against the Celestia sanity limit.
+					select! {
+						res = pending_request => {
+							let (block_height, celestia_height) = res?;
+							if self
+								.notifier
+								.send(
+									ExternalDaNotification::BlockCommitted(block_height, celestia_height)
```

Comment:
The block has been submitted not committed. We need to wait for the Celestia to produce the blob.

Reply:
Isn't this what [celestia RPC method](https://node-rpc-docs.celestia.org/?version=v0.21.9#blob.Submit) does?

Reply:
Yes, I didn't know. I thought you need to submit and then pull until you get the height like it's done in the current client. Ok, so it's fine.
Can you look in the current client why we've changed, the height pulling to pull certificate then pull height? Perhaps we need to pull a proof with `GetProof` or something like that?

Reply:
In the current architecture, submitting blobs and getting blobs for the current height is done in separate tasks. There may be more than one node submitting blobs, and Celestia serves as the sequencer.